compiler/hash-parser/src/lexer.rs

```diff
@@ -0,0 +1,618 @@
+        }
+    }
+
+    /// Consume an exponent for a float literal
+    fn eat_exponent(&self) -> TokenResult<i32> {
+        debug_assert!(matches!(self.prev.get(), 'e' | 'E'));
+        let start = self.offset.get();
+
+        // Check if there is a sign before the digits start in the exponent...
+        let negated = if self.peek() == '-' {
+            self.next();
+            true
+        } else {
+            false
+        };
+
+        let num = self.eat_decimal_digits().collect::<String>().parse::<i32>();
+
+        // Ensure that the digit parsing was ok
+        if num.is_err() {
+            return Err(TokenError::new(
+                Some("Invalid float exponent.".to_string()),
+                TokenErrorKind::MalformedNumericalLiteral,
+                Location::span(start, self.offset.get()),
+            ));
+        }
+
+        let num = num.unwrap();
+
+        if negated {
+            Ok(-num)
+        } else {
+            Ok(num)
+        }
+    }
+
+    /// Consume only decimal digits up to encountering a non-decimal digit
+    /// whilst taking into account that the language supports '_' as digit
+    /// separators which should just be skipped over...
+    pub(crate) fn eat_decimal_digits(&self) -> impl Iterator<Item = char> + '_ {
+        self.eat_while(|peeked| matches!(peeked, '0'..='9' | '_'))
+            .filter(|e| *e != '_')
+    }
+
+    /// Transform an ordinary character into a well known escape sequence specified by the
+    /// escape literal rules. More information about the escape sequences can be found at
+    /// [escape sequences](https://hash-org.github.io/lang/basics/intro.html)
+    fn char_from_escape_seq(&self) -> TokenResult<char> {
+        debug_assert!(self.prev.get() == '\\');
+
+        // @@Incomplete: come up with a better algorithm to transform escaped literals, rather than manual
+        // transformations!
+        let c = self.next().unwrap();
+
+        // we need to compute the old byte offset by accounting for both the 'u' character and the '\\' character,
+        // but since this is known to be 2 bytes, we can just subtract it from the current offset
+        let start = self.offset.get() - 1;
+
+        match c {
+            'n' => Ok('\n'),
+            't' => Ok('\t'),
+            'u' => {
+                // The next character should be a '{', otherwise this isn't a correct escaped
+                // literal
+                if self.peek() != '{' {
+                    return Err(TokenError::new(
+                        Some("Expected '{' after a '\\u' escape sequence".to_string()),
+                        TokenErrorKind::BadEscapeSequence,
+                        Location::span(start, self.offset.get()),
+                    ));
+                }
+
+                self.next();
+
+                // here we expect up to 6 hex digits, which is finally closed by a '}'
+                let chars: String = self.eat_while(|c| c.is_ascii_hexdigit()).collect();
```

Comment:
Is it necessary to collect here? Maybe could use the slice from the input?

Reply:
In any case, it still has to be collected into a string since the `parse` method expects one.